src/particles/pusher/FieldGather.H

```diff
@@ -204,4 +204,249 @@ void doGatherShapeN (const amrex::ParticleReal xp,
+ */
+template <int depos_order_xy, int depos_order_z>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+void doLaserGatherShapeN (const amrex::ParticleReal xp,
+                          const amrex::ParticleReal yp,
+                          const amrex::ParticleReal zp,
+                          amrex::ParticleReal& Aabssqp,
+                          amrex::ParticleReal& AabssqDxp,
+                          amrex::ParticleReal& AabssqDyp,
+                          amrex::Array4<amrex::Real const> const& aabssq_arr,
+                          amrex::Array4<amrex::Real const> const& aabssqdx_arr,
+                          amrex::Array4<amrex::Real const> const& aabssqdy_arr,
+                          const amrex::GpuArray<amrex::Real, 3>& dx,
+                          const amrex::Real x_pos_offset,
+                          const amrex::Real y_pos_offset,
+                          const amrex::Real z_pos_offset)
+{
+    using namespace amrex::literals;
+
+    // x,y,z direction
+    const amrex::Real x = (xp-x_pos_offset)/dx[0];
+    const amrex::Real y = (yp-y_pos_offset)/dx[1];
+    const amrex::Real z = (zp-z_pos_offset)/dx[2];
+
+    // --- Compute shape factors
+    // x direction
+    // j_cell leftmost cell in x that the particle touches. sx_cell shape factor along x
+    amrex::Real sx_cell[depos_order_xy + 1];
+    const int j_cell = compute_shape_factor<depos_order_xy>(sx_cell, x);
+
+    // y direction
+    amrex::Real sy_cell[depos_order_xy + 1];
+    const int k_cell = compute_shape_factor<depos_order_xy>(sy_cell, y);
+
+    // z direction;
+    amrex::Real sz_cell[depos_order_z + 1];
+    const int l_cell = compute_shape_factor<depos_order_z>(sz_cell, z);
+    amrex::ignore_unused(l_cell);
+
+    // Gather field on particle Aabssqp from field on grid aabssq_arr
+    for (int iz=0; iz<=depos_order_z; iz++){
+        for (int iy=0; iy<=depos_order_xy; iy++){
+            for (int ix=0; ix<=depos_order_xy; ix++){
+                Aabssqp += sx_cell[ix]*sy_cell[iy]*sz_cell[iz]*
+                    aabssq_arr(j_cell+ix, k_cell+iy, amrex::lbound(aabssq_arr).z);
+            }
+        }
+    }
+    // Gather field on particle AabssqDxp from field on grid aabssqdx_arr
+    for (int iz=0; iz<=depos_order_z; iz++){
+        for (int iy=0; iy<=depos_order_xy; iy++){
+            for (int ix=0; ix<=depos_order_xy; ix++){
+                AabssqDxp += sx_cell[ix]*sy_cell[iy]*sz_cell[iz]*
+                    aabssqdx_arr(j_cell+ix, k_cell+iy, amrex::lbound(aabssqdx_arr).z);
+            }
+        }
+    }
+    // Gather field on particle AabssqDyp from field on grid aabssqdy_arr
+    for (int iz=0; iz<=depos_order_z; iz++){
+        for (int iy=0; iy<=depos_order_xy; iy++){
+            for (int ix=0; ix<=depos_order_xy; ix++){
+                AabssqDyp += sx_cell[ix]*sy_cell[iy]*sz_cell[iz]*
+                    aabssqdy_arr(j_cell+ix, k_cell+iy, amrex::lbound(aabssqdy_arr).z);
+            }
+        }
+    }
+}
```

Comment:
As discussed offline, we can probably NOT store the transverse derivative and always compute them on-the-fly. Then, these 3 blocks could be merged, should be faster and more memory-efficient.